src/app/clusters/service-area-server/service-area-server.h

```diff
@@ -0,0 +1,751 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "service-area-cluster-objects.h"
+#include <app-common/zap-generated/cluster-objects.h>
+
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/ConcreteCommandPath.h>
+#include <app/util/af-types.h>
+
+#include <app/util/basic-types.h>
+#include <app/util/config.h>
+#include <lib/support/Span.h>
+#include <platform/CHIPDeviceConfig.h>
+
+#include <list>
+#include <map>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ServiceArea {
+
+
+// *****************************************************************************
+// cluster constraints
+
+constexpr size_t   kMaxNumSupportedLocations  = 255;
+constexpr size_t   kMaxNumSupportedMaps       = 255;
+constexpr size_t   kMaxNumSelectedLocations   = 255;
+constexpr size_t   kMaxNumSupportedProgress   = 255;
+
+
+/**
+ * Uncopyable is a helper to ensure that Service Area cluster instances and delegates
+ * are unique and cannot be copied.
+ */
+class Uncopyable
+{
+protected:
+    Uncopyable() {}
+    ~Uncopyable() = default;
+
+private:
+    Uncopyable(const Uncopyable &)             = delete;
+    Uncopyable & operator=(const Uncopyable &) = delete;
+};
+
+
+class Delegate;
+
+
+/**
+ * Instance is a class that represents an instance of the generic Service Area cluster.
+ * It implements AttributeAccessInterface and CommandHandlerInterface so it can
+ * handle commands for any implementation of the location cluster.
+ * Custom implementations of the Service Area cluster override functions in the Delegate class
+ * must be provided to operate with specific devices.
+ */
+class Instance : public AttributeAccessInterface, public CommandHandlerInterface, public Uncopyable
+{
+public:
+    /**
+     * @brief Creates a Service Area cluster instance. The Init() function needs to be called for this instance
+     * to be registered and called by the interaction model at the appropriate times.
+     * @param aDelegate A pointer to the delegate to be used by this server.
+     * Note: the caller must ensure that the delegate lives throughout the instance's lifetime.
+     * @param aEndpointId The endpoint on which this cluster exists. This must match the zap configuration.
+     * @param aFeature The supported features of this Service Area Cluster
+     * @param aClusterId The Id of the Service Area cluster to be instantiated.
+     */
+    Instance(Delegate * aDelegate, EndpointId aEndpointId, BitMask<ServiceArea::Feature> aFeature, ClusterId aClusterId = Clusters::ServiceArea::Id);
+
```

Comment:
In some of the other server implementations they like being able to pass in the cluster Id. Not sure why you would want to do this, but it is more flexible. I kept the option, but provided the default value to you don't have to pass it in unless you want to.